modules/navier_stokes/src/fvkernels/PCNSFVKT.C

```diff
@@ -0,0 +1,269 @@
+{
+  if ((_eqn == "momentum") && !isParamValid("momentum_component"))
+    paramError("eqn",
+               "If 'momentum' is specified for 'eqn', then you must provide a parameter "
+               "value for 'momentum_component'");
+
+#ifndef MOOSE_GLOBAL_AD_INDEXING
+  mooseError(
+      "PCNSFVKT objects require global AD indexing because they  require gradient "
+      "reconstruction. Please cd into the root of your MOOSE directory and run './configure'");
+#endif
+}
+
+std::pair<ADReal, ADReal>
+PCNSFVKT::computeAlphaAndOmega(const ADReal & u_elem_normal,
+                               const ADReal & u_neighbor_normal,
+                               const ADReal & c_elem,
+                               const ADReal & c_neighbor) const
+{
+  // Equations from Greenshields sans multiplication by area (which well be done in
+  // computeResidual/Jacobian
+  const auto psi_elem =
+      std::max({c_elem + u_elem_normal, c_neighbor + u_neighbor_normal, ADReal(0)});
+  const auto psi_neighbor =
+      std::max({c_elem - u_elem_normal, c_neighbor - u_neighbor_normal, ADReal(0)});
+  auto alpha = _knp_for_omega ? psi_elem / (psi_elem + psi_neighbor) : ADReal(0.5);
+  auto omega = _knp_for_omega ? alpha * (1 - alpha) * (psi_elem + psi_neighbor)
+                              : alpha * std::max(psi_elem, psi_neighbor);
+
+  // Do this to avoid new nonzero mallocs
+  const auto dummy_quant = 0 * (c_elem + u_elem_normal + c_neighbor + u_neighbor_normal);
+
+  alpha += dummy_quant;
+  omega += dummy_quant;
+  return std::make_pair(std::move(alpha), std::move(omega));
+}
+
+ADReal
+PCNSFVKT::computeFaceFlux(const ADReal & alpha,
+                          const ADReal & omega,
+                          const ADReal & sup_vel_elem_normal,
+                          const ADReal & sup_vel_neighbor_normal,
+                          const ADReal & adv_quant_elem,
+                          const ADReal & adv_quant_neighbor)
+{
+  return alpha * (sup_vel_elem_normal * adv_quant_elem) +
+         (1 - alpha) * sup_vel_neighbor_normal * adv_quant_neighbor -
+         omega * (adv_quant_neighbor - adv_quant_elem);
+}
+
+ADReal
+PCNSFVKT::computeQpResidual()
+{
+  // Perform primitive interpolations
+  const auto pressure_elem = interpolate(*_limiter,
+                                         _pressure_elem[_qp],
+                                         _pressure_neighbor[_qp],
+                                         _grad_pressure_elem[_qp],
+                                         *_face_info,
+                                         /*elem_is_up=*/true);
+  const auto pressure_neighbor = interpolate(*_limiter,
+                                             _pressure_neighbor[_qp],
+                                             _pressure_elem[_qp],
+                                             _grad_pressure_neighbor[_qp],
+                                             *_face_info,
+                                             /*elem_is_up=*/false);
```

Comment:
greenshields interpolates rho directly

Reply:
That's true. In the article, he interpolates a mixed variable set: density, momentum, and temperature. However, in their code `rhoCentralFoam` they actually interpolate density, momentum, specific internal energy, and 1/compressibility. It's somewhat interesting that a variable for which they express concerns about boundedness is a variable that they don't interpolate in the code.

I think we are free to choose the variables that we interpolate. With the current choice of interpolants I have not seen any variables that fall outside physical bounds.

Reply:
I like the primitive variable set because it fits best with what users typically specify boundary conditions for.